securedrop_client/gui/widgets.py

```diff
@@ -789,52 +789,57 @@ def update(self, sources: List[Source]) -> List[str]:
         """
         Update the list with the passed in list of sources.
         """
-        # Delete widgets that no longer exist in source list
-        source_uuids = [source.uuid for source in sources]
+        sources_to_update = []
+        sources_to_add = []
+        for source in sources:
+            try:
+                if source.uuid in self.source_items:
+                    sources_to_update.append(source.uuid)
+                else:
+                    sources_to_add.append(source)
+            except sqlalchemy.exc.InvalidRequestError as e:
+                logger.debug(e)
+                continue
+
+        # Delete widgets for sources not in the supplied sourcelist
         deleted_uuids = []
-        for i in range(self.count()):
-            list_item = self.item(i)
-            list_widget = self.itemWidget(list_item)
+        sources_to_delete = [self.source_items[uuid] for uuid in self.source_items
+                             if uuid not in sources_to_update]
+        for source_item in sources_to_delete:
+            if source_item.isSelected():
+                self.setCurrentItem(None)
 
-            if list_widget and list_widget.source_uuid not in source_uuids:
-                if list_item.isSelected():
-                    self.setCurrentItem(None)
+            source_widget = self.itemWidget(source_item)
+            self.takeItem(self.row(source_item))
+            source_widget.deleteLater()
```

Comment:
I noticed during running through the STR for #1117 that there are lingering blank source list items (removed after subsequent run of the client, but remain after subsequent syncs):

<img width="573" alt="Screen Shot 2020-06-17 at 5 03 36 PM" src="https://user-images.githubusercontent.com/7832803/84950592-25573080-b0bd-11ea-8e56-31dc8fe90705.png">

In the debug logs there _are_ a bunch of entries like:

```
2020-06-17 17:06:05,141 - securedrop_client.gui.widgets:1065(update) DEBUG: Could not update SourceWidget for source 32803f81-95b2-4a3d-bb35-6106fa7719a5: Could not refresh instance '<Source at 0x11538ad10>'
```

indicating that the added exception handling is preventing the crash, but the items are not getting cleaned up.

Reply:
I was able to repro your finding and have tested your solution posted https://github.com/freedomofpress/securedrop-client/commit/5d06b0bfd94d7cffc762607b64edcc6c962341dc and it seems to work! It also makes sense because now we make sure adding to the cache can't raise so that we ensure all created source widgets are added to the cache.  Thinking of how to make sure we cover this in a regression test.

Reply:
As followup I added a regression test that raises an exception when you try to access uuid on the source object, and made sure that _both_ the cache and the SourceList are empty after calling initial_update: https://github.com/freedomofpress/securedrop-client/pull/1114/commits/97b3d3d6e4029c64e106d5cc21e1d733c78c8159